E2E test: Connection resilience — operations after socket error

## Problem

No E2E test verifies client behavior when the **underlying TCP socket encounters an error** mid-session. In production, network partitions, server restarts, and load balancer timeouts cause socket errors.

## Expected tests

```php
public function testOperationAfterSocketDisconnect(): void
{
    $connection = Connection::create(...);
    $streamName = 'test-resilience-' . uniqid();
    $connection->createStream($streamName);

    // Simulate socket error by stopping Docker container
    // or closing the socket via reflection

    $this->expectException(\RuntimeException::class);
    $connection->createStream('another-stream');
}

public function testIsConnectedReturnsFalseAfterSocketError(): void
{
    // After socket error, isConnected() should return false
}
```

## Why it matters

- Network failures are inevitable in production
- The client must fail fast with clear errors, not hang or corrupt state
- `readBytes()` returns `null` on `SOCKET_ETIMEDOUT` but throws on other errors — both paths need testing

## Notes

- This test may require Docker manipulation (stop/start container)
- Alternative: use reflection to close the socket directly
- Consider marking as `@group slow` or `@group destructive`

## Acceptance criteria

- [ ] Socket error → clear `RuntimeException` on next operation
- [ ] `isConnected()` returns `false` after socket error
- [ ] No resource leaks after socket error

## CI Pipeline Failed

Pipeline checks failed. Logs saved to `/home/decodo/work/rabbit-stream/.oda/artifacts/170/pipeline-fail.log`.

```
## Failed CI Checks

Failed: Lint (PHPCS + Rector + PHPStan)

### Run 23745087081

```
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	﻿2026-03-30T12:38:44.1780438Z ##[group]Run composer rector
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T12:38:44.1780962Z [36;1mcomposer rector[0m
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T12:38:44.1806994Z shell: /usr/bin/bash -e {0}
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T12:38:44.1807227Z env:
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T12:38:44.1807417Z   COMPOSER_PROCESS_TIMEOUT: 0
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T12:38:44.1807656Z   COMPOSER_NO_INTERACTION: 1
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T12:38:44.1807875Z   COMPOSER_NO_AUDIT: 1
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T12:38:44.1808080Z ##[endgroup]
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T12:38:52.7227186Z    0/212 [░░░░░░░░░░░░░░░░░░░░░░░░░░░░]   0%
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T12:38:52.8110176Z 
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T12:38:52.8110699Z 1 file with changes
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T12:38:52.8111525Z ===================
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T12:38:52.8111839Z 
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T12:38:52.8115014Z 1) src/StreamConnection.php:118
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T12:38:52.8115581Z 
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T12:38:52.8118329Z     ---------- begin diff ----------
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T12:38:52.8118803Z @@ @@
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T12:38:52.8119064Z                  $this->connected = false;
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T12:38:52.8119424Z                  return false;
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T12:38:52.8119719Z              }
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T12:38:52.8119967Z -        } catch (\Error $e) {
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T12:38:52.8120341Z +        } catch (\Error) {
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T12:38:52.8120688Z              // Socket is invalid/closed
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T12:38:52.8121471Z              $this->connected = false;
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T12:38:52.8122019Z              return false;
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T12:38:52.8122427Z @@ @@
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T12:38:52.8122772Z              $written = socket_write($this->socket, $frame, strlen($frame));
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T12:38:52.8123268Z          } catch (\Error $e) {
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T12:38:52.8123576Z              $this->connected = false;
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T12:38:52.8124143Z -            throw new ConnectionException("Failed to write to socket: " . $e->getMessage());
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T12:38:52.8125544Z +            throw new ConnectionException("Failed to write to socket: " . $e->getMessage(), $e->getCode(), $e);
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T12:38:52.8126281Z          }
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T12:38:52.8126416Z 
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T12:38:52.8126534Z          if ($written === false) {
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T12:38:52.8126873Z     ----------- end diff -----------
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T12:38:52.8127069Z 
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T12:38:52.8127155Z Applied rules:
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T12:38:52.8127361Z  * ThrowWithPreviousExceptionRector
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T12:38:52.8127678Z  * ClosureFromCallableToFirstClassCallableRector
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T12:38:52.8127987Z  * RemoveUnusedVariableInCatchRector
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T12:38:52.8128163Z 
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T12:38:52.8128167Z 
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T12:38:52.8142485Z  [OK] 1 file would have been changed (dry-run) by Rector                        
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T12:38:52.8142899Z 
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T12:38:52.8254764Z  212/212 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓] 100%Script rector process --dry-run handling the rector event returned with error code 2
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T12:38:52.8359849Z ##[error]Process completed with exit code 2.
```


```

--- a/src/Client/Connection.php
+++ b/src/Client/Connection.php
@@ -262,6 +262,11 @@ class Connection implements ConnectionInterface
         }
     }
 
+    public function isConnected(): bool
+    {
+        return $this->streamConnection->isConnected();
+    }
+
     public function __destruct()
     {
         if (!$this->closed) {

--- a/src/StreamConnection.php
+++ b/src/StreamConnection.php
@@ -118,7 +118,7 @@ class StreamConnection
                 $this->connected = false;
                 return false;
             }
-        } catch (\Error $e) {
+        } catch (\Error) {
             // Socket is invalid/closed
             $this->connected = false;
             return false;
@@ -243,7 +243,7 @@ class StreamConnection
             $written = socket_write($this->socket, $frame, strlen($frame));
         } catch (\Error $e) {
             $this->connected = false;
-            throw new ConnectionException("Failed to write to socket: " . $e->getMessage());
+            throw new ConnectionException("Failed to write to socket: " . $e->getMessage(), $e->getCode(), $e);
         }
 
         if ($written === false) {

--- a/src/StreamConnection.php
+++ b/src/StreamConnection.php
@@ -101,7 +101,30 @@ class StreamConnection
 
     public function isConnected(): bool
     {
-        return $this->connected;
+        if (!$this->connected) {
+            return false;
+        }
+
+        if (!$this->socket instanceof \Socket) {
+            return false;
+        }
+
+        // Check if socket is still valid by attempting to get its error status
+        // A closed socket will fail this operation
+        try {
+            $error = @socket_last_error($this->socket);
+            if ($error !== 0) {
+                // Socket has an error, mark as disconnected
+                $this->connected = false;
+                return false;
+            }
+        } catch (\Error $e) {
+            // Socket is invalid/closed
+            $this->connected = false;
+            return false;
+        }
+
+        return true;
     }
 
     public function setMaxFrameSize(int $maxFrameSize): void
@@ -216,7 +239,13 @@ class StreamConnection
             throw new ConnectionException("Cannot write: socket is not connected");
         }
 
-        $written = socket_write($this->socket, $frame, strlen($frame));
+        try {
+            $written = socket_write($this->socket, $frame, strlen($frame));
+        } catch (\Error $e) {
+            $this->connected = false;
+            throw new ConnectionException("Failed to write to socket: " . $e->getMessage());
+        }
+
         if ($written === false) {
             throw new ConnectionException(
                 "Failed to write to socket: " . socket_strerror(socket_last_error($this->socket))

--- a/tests/E2E/ConnectionResilienceTest.php
+++ b/tests/E2E/ConnectionResilienceTest.php
@@ -1,0 +1,107 @@
+<?php
+
+declare(strict_types=1);
+
+namespace CrazyGoat\RabbitStream\Tests\E2E;
+
+use CrazyGoat\RabbitStream\Client\Connection;
+use CrazyGoat\RabbitStream\Exception\ConnectionException;
+use CrazyGoat\RabbitStream\StreamConnection;
+use PHPUnit\Framework\TestCase;
+
+/**
+ * @group destructive
+ * @group slow
+ */
+class ConnectionResilienceTest extends TestCase
+{
+    private static string $host = '127.0.0.1';
+    private static int $port = 5552;
+
+    public static function setUpBeforeClass(): void
+    {
+        self::$host = getenv('RABBITMQ_HOST') ?: self::$host;
+        self::$port = (int)(getenv('RABBITMQ_PORT') ?: self::$port);
+    }
+
+    public function testOperationAfterSocketDisconnect(): void
+    {
+        $connection = Connection::create(
+            host: self::$host,
+            port: self::$port,
+            user: 'guest',
+            password: 'guest',
+            vhost: '/'
+        );
+
+        $streamName = 'test-resilience-' . uniqid();
+        $connection->createStream($streamName);
+
+        // Simulate socket error by closing socket via reflection
+        $this->forceCloseSocket($connection);
+
+        // Next operation should throw ConnectionException
+        $this->expectException(ConnectionException::class);
+        $connection->createStream('another-stream-' . uniqid());
+    }
+
+    public function testIsConnectedReturnsFalseAfterSocketError(): void
+    {
+        $connection = Connection::create(
+            host: self::$host,
+            port: self::$port,
+            user: 'guest',
+            password: 'guest',
+            vhost: '/'
+        );
+
+        $this->assertTrue($connection->isConnected());
+
+        // Simulate socket error
+        $this->forceCloseSocket($connection);
+
+        // isConnected() should return false
+        $this->assertFalse($connection->isConnected());
+    }
+
+    public function testNoResourceLeaksAfterSocketError(): void
+    {
+        $connection = Connection::create(
+            host: self::$host,
+            port: self::$port,
+            user: 'guest',
+            password: 'guest',
+            vhost: '/'
+        );
+
+        $streamName = 'test-resilience-leak-' . uniqid();
+        $connection->createStream($streamName);
+
+        // Force socket close
+        $this->forceCloseSocket($connection);
+
+        // Destructor should not throw or cause warnings
+        unset($connection);
+
+        // If we reach here without errors, test passes
+        $this->addToAssertionCount(1);
+    }
+
+    private function forceCloseSocket(Connection $connection): void
+    {
+        // Access StreamConnection from Connection
+        $streamConnReflection = new \ReflectionProperty(Connection::class, 'streamConnection');
+        $streamConnection = $streamConnReflection->getValue($connection);
+
+        $this->assertInstanceOf(StreamConnection::class, $streamConnection);
+
+        // Access socket from StreamConnection
+        $socketReflection = new \ReflectionProperty(StreamConnection::class, 'socket');
+        $socket = $socketReflection->getValue($streamConnection);
+
+        if ($socket !== null) {
+            // Force close the socket
+            @socket_close($socket);
+        }
+    }
+}

--- a/tests/E2E/ConnectionResilienceTest.php
+++ b/tests/E2E/ConnectionResilienceTest.php
@@ -99,7 +99,7 @@ class ConnectionResilienceTest extends TestCase
         $socketReflection = new \ReflectionProperty(StreamConnection::class, 'socket');
         $socket = $socketReflection->getValue($streamConnection);
 
-        if ($socket !== null) {
+        if ($socket instanceof \Socket) {
             // Force close the socket
             @socket_close($socket);
         }

